app/services/segment_delete_service.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+class SegmentDeleteService
+  SEGMENT_URI = URI('https://gdpr.segment.com/graphql')
+
+  class SegmentDeleteServiceError < StandardError
+    def initialize(message:, response:)
+      super(message)
+      @response = {body: response.body, code: response.code}
+    end
+
+    attr_reader :response
+
+    def error_type
+      :delete_service
+    end
+  end
+  class SegmentAuthenticationError < SegmentDeleteServiceError
+    def initialize(message: 'Cannot authenticate', response:)
+      super(message: message, response: response)
+    end
+
+    def error_type
+      :authentication
+    end
+  end
+  class SegmentUserDeletionError < SegmentDeleteServiceError
+    def initialize(message: 'Cannot delete user', response:)
+      super(message: message, response: response)
+    end
+
+    def error_type
+      :user_deletion
+    end
+  end
+
+
+  def initialize(user_id)
+    @user_id = user_id
+    @errors = ActiveModel::Errors.new(self)
+  end
+
+  attr_reader :user_id, :errors
+
+  def call!
+    return true unless Features::SegmentDeletionConfig.enabled?
+    request_body = {query: "mutation { createWorkspaceRegulation(workspaceSlug: \"3scale\" type: SUPPRESS_AND_DELETE userId: \"#{user_id}\" ) { id } }"}.to_json
+    response = perform_request(request_body: request_body, authorization: request_token)
+    raise SegmentUserDeletionError.new(response: response) unless response.code.to_s == '200'
+    :ok
+  end
+
+  def call
+    call!
+  rescue SegmentDeleteServiceError => exception
```

Comment:
Oh wait you raise in `#call!` and rescue here?
Do we really need `#call!` ? Could we avoid raising exceptions and catching them in the same class?
I think exceptions are not meant to be used like that. They are _exceptions_, meaning they should be handled particularly when used outside of the class not inside.

By the way it is the other way around:

```ruby
def save
  safely_save_in_database
end

def save!
  save or raise SaveError
end
```

Reply:
`call!` with the bang is currently used to be rescue from the worker or to raise in the worker, and `call` without the bang is not being called from anywhere. So the main purpose of these exceptions weren't for the same class but for the worker.
Thanks for the order they are supposed to be anyway 👍 I will keep it mind when I need it 👍 
Not like this comment of mine really matters much now since the whole code will change a loooot anyway, but I prefer to leave it clear 😄